src/FastFeatureVector/FeatureVector.scala

```diff
@@ -194,21 +199,39 @@ case class FeatureVector(labelset : Array[String],
         return f
     } */
     def read(iterator: Iterator[String]) {
-        val regex = ("""(.*?)(\+L=("""+labelset.mkString("|")+"""))?[ \t]([^ \t]*)""").r  // .*? is non-greedy
-        // (feature, _, label, value)
-        // matches featurename+L=label 1.0
-        fmap.clear()
-        for (line <- iterator) {
-            val regex(feature, _, label, value) = line
-            if (!fmap.contains(feature)) {
-                fmap(feature) = ValuesMap(0.0, Map())
-            }
-            if (label == null) {
-                fmap(feature).unconjoined = value.toDouble
-            } else {
-                fmap(feature).conjoined(labelToIndex(label)) = value.toDouble // TODO: catch invalid label errors and print labels
-            }
+      val newRegex = ( """(.*?)(\+L=(\:[a-zA-Z0-9-]+))?[ \t]([^ \t]*)""" ).r // .*? is non-greedy
```

Comment:
Awesome. Here’s my
.config/IntelliJIdea/14/config/codestyles/JeffScalaStyle.xml. looks like
tab/indent size is the only idiosyncrasy:

<?xml version="1.0" encoding="UTF-8"?>
<code_scheme name="JeffScalaStyle">
  <XML>
    <option name="XML_LEGACY_SETTINGS_IMPORTED" value="true" />
  </XML>
  <codeStyleSettings language="Scala">
    <indentOptions>
      <option name="INDENT_SIZE" value="4" />
      <option name="CONTINUATION_INDENT_SIZE" value="4" />
      <option name="TAB_SIZE" value="4" />
    </indentOptions>
  </codeStyleSettings>
</code_scheme>

​

On Wed, Jan 14, 2015 at 1:12 PM, peteoleary <notifications@github.com>
wrote:

> In src/FastFeatureVector/FeatureVector.scala
> <https://github.com/jflanigan/jamr/pull/8#discussion-diff-22956437>:
>
> > @@ -194,21 +199,39 @@ case class FeatureVector(labelset : Array[String],
> >          return f
> >      } */
> >      def read(iterator: Iterator[String]) {
> > -        val regex = ("""(.*?)(\+L=("""+labelset.mkString("|")+"""))?[ \t]([^ \t]*)""").r  // .*? is non-greedy
> > -        // (feature, _, label, value)
> > -        // matches featurename+L=label 1.0
> > -        fmap.clear()
> > -        for (line <- iterator) {
> > -            val regex(feature, _, label, value) = line
> > -            if (!fmap.contains(feature)) {
> > -                fmap(feature) = ValuesMap(0.0, Map())
> > -            }
> > -            if (label == null) {
> > -                fmap(feature).unconjoined = value.toDouble
> > -            } else {
> > -                fmap(feature).conjoined(labelToIndex(label)) = value.toDouble // TODO: catch invalid label errors and print labels
> > -            }
> > +      val newRegex = ( """(.*?)(\+L=(\:[a-zA-Z0-9-]+))?[ \t]([^ \t]*)""" ).r // .*? is non-greedy
>
> I'm using IntelliJ and had a custom Code Style selected so that's what's
> messing up the whitespace. I've set it it 4 space indents now so hopefully
> it will cause less churn in the future.
>
> What I've done here is replace regex with a simpler (and faster) newRegex
> which does not hardcode the labelset in it. This has the effect of letting
> unknown labels through the parse; I then catch these unknown labels and
> print them all afterwards.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jflanigan/jamr/pull/8/files#r22956437>.
>